File: 57-insert-interval/insert-interval.py
# Brute Force Code
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        # add the new interval to the existing list
        intervals.append(newInterval)
        # sort all intervals by their starting points
        intervals.sort()
        # store the final merged intervals
        result = []
        # traverse the every interval
        for interval in intervals:
            # if this is the first interval add its directly
            if len(result) == 0:
                result.append(interval)
            else:
                # get the last merged interval
                last_interval = result[-1]
                # check whether the current interval overlap with the last merged interval
                if interval[0] <= last_interval[1]:
                    # merge the intervals by extending the ending point
                    last_interval[1] = max(last_interval[1], interval[1]
                    )
                else:
                    # no overlap start new interval
                    result.append(interval)
        # return the merged interval
        return result 

# Time Complexity : O(Nlog)
# Space Complxity : O(1)

File: 57-insert-interval/insert-interval.cpp
// Brute Force Code & Optimal Code
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        // add the new interval to the existing list
        intervals.push_back(newInterval);
        // sort all intervals by their starting points
        sort(intervals.begin(), intervals.end());
        // store the final merged intervals
        vector<vector<int>> result;
        // traverse every interval
        for (vector<int>& interval : intervals) {
            // if this is the first interval add it directly
            if (result.empty()) {
                result.push_back(interval);
            }
            else {
                // get the last merged interval
                vector<int>& last_interval = result.back();
                // check whether the current interval overlaps with the last merged interval
                if (interval[0] <= last_interval[1]) {
                    // merge the intervals by extending the ending point
                    last_interval[1] = max(last_interval[1], interval[1]);
                }
                else {
                    // no overlap, start a new interval
                    result.push_back(interval);
                }
            }
        }
        // return the merged intervals
        return result;
    }
};

// Time Complexity : O(Nlog)
// Space Complexity : O(1)